includes/page-controller/class-wc-admin-page-controller.php

```diff
@@ -0,0 +1,398 @@
+<?php
+/**
+ * WC_Admin_Page_Controller
+ *
+ * @package Woocommerce Admin
+ */
+
+/**
+ * WC_Admin_Page_Controller
+ */
+class WC_Admin_Page_Controller {
+	// JS-powered page root.
+	const PAGE_ROOT = 'wc-admin';
+
+	/**
+	 * Singleton instance of self.
+	 *
+	 * @var WC_Admin_Page_Controller
+	 */
+	private static $instance = false;
+
+	/**
+	 * Current page ID (or false if not registered with this controller).
+	 *
+	 * @var string
+	 */
+	private $current_page = null;
+
+	/**
+	 * Registered pages
+	 * Contains information (breadcrumbs, menu info) about JS powered pages and classic WooCommerce pages.
+	 *
+	 * @var array
+	 */
+	private $pages = array();
+
+	/**
+	 * We want a single instance of this class so we can accurately track registered menus and pages.
+	 */
+	public static function get_instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Connect an existing page to wc-admin.
+	 *
+	 * @param array $options {
+	 *   Array describing the page.
+	 *
+	 *   @type string       id           Id to reference the page.
+	 *   @type string|array title        Page title. Used in menus and breadcrumbs.
+	 *   @type string|null  parent       Parent ID. Null for new top level page.
+	 *   @type string       path         Path for this page. E.g. admin.php?page=wc-settings&tab=checkout
+	 *   @type string       capability   Capability needed to access the page.
+	 *   @type string       icon         Icon. Dashicons helper class, base64-encoded SVG, or 'none'.
+	 *   @type int          position     Menu item position.
+	 * }
+	 */
+	public function connect_page( $options ) {
+		if ( ! is_array( $options['title'] ) ) {
+			$options['title'] = array( $options['title'] );
+		}
+
+		// TODO: check for null ID, or collision.
+		$this->pages[ $options['id'] ] = $options;
```

Comment:
I'm wondering if it might be worth running `$options` through a filter before adding it here?

The use case I'm thinking of here is the eCommerce Plan, where menu/submenu information is passed in a similar manner: https://github.com/Automattic/wc-calypso-bridge/blob/master/includes/connect/woocommerce-bookings.php In the future, those connect functions could be re-written to append extra info to these existing arrays.